### PR TITLE
Added test cases & docs

### LINKED ISSUE #1: Add tests
Need to create tests of the math functionality in this crate.

### PR DESCRIPTION
**Pull request by RoboRambo**
Resolves Issues #1

Added Test cases to cover more edge cases for inputs of all functions
Added Docs

### CODE REVIEW

**Comment by RoboRambo** [comment]
Hello there.

I hope that at least some of the Test Cases make sense.
I am not sure what to add to the Docs of fn calculate_liquidity_in & calculate_luquidity_out though.
I do have a few more thoughts on this (e.g.: rounding of the result of fn calculate_spot_price), I will can put those to words once I get some sleep.

Any Feedback is appreciated.

**Comment by RoboRambo** [comment]
1. Rounding

It seems that the functions round their results as follows:

{calculate_spot_price, calculate_liquidity_in, calculate_liquidity_out} - truncate
{calculate_sell_price, calculate_buy_price} - round up

I guess this is as designed, however let me state one corner case for fn calculate_spot_price:
Formula: BUY_RESERVE * AMOUNT / SELL_RESERVE
Input:
BUY_RESERVE = u128::MAX - 1
SELL_RESERVE = u128::MAX
AMOUNT = 1

u128::MAX-1 * 1 / u128::MAX = 0,9999

Output:
0

Maybe this is of no concern. If it is however, I can think of some more corner cases.

2. Exception Handling

It is quite easy for a Client to use the functions in a "erronous" manner.

Using fn calculate_spot_price with the following input (which should be semantically correct) might return an unexpected result:
Input:
BUY_RESERVE = sqrt(u128::MAX) + 1
SELL_RESERVE = sqrt(u128::MAX) + 1
AMOUNT = 1

Output:
None

Also, a Client will get the same result for the following input (which is semantically incorrect/formula div by zero):
Input:
BUY_RESERVE = any
SELL_RESERVE = 0
AMOUNT = any

Output:
None


Maybe it would be worth the effort to try to distinguish between these cases, as I think that a Client should not have to know implementation details of a function in order to be able to use it properly.


However I am new to Rust/Web3 and maybe all of this makes sense in given the Context.
Feedback would be appreciated. :)

Thanks in advance

**Comment by enthusiastmartin** [comment]
Thanks. 

Yes, absolutely correct. However it is work in progress as this math was extracted as is from the main repo where such bad cases would never happen ( like sell_reserve = 0 ).

But there cannot be such assumptions here of course.

Would you be willing to look at property testing - something like this [proptest](https://altsysrq.github.io/proptest-book/intro.html) and use it here if possible ? 

Another issue is to use Result for proper error handling ( #3  ).

**Comment by RoboRambo** [comment]
Thanks for the reply, I will surely look into property testing.

Had a conversation regarding issue today with jakub, need to process all the data :)

I do have one question though:
Docs for fn calculate_buy_price state:
`/// Calculating buying price given reserve of selling asset and reserve of buying asset.
/// Formula : SELL_RESERVE * AMOUNT / (BUY_RESERVE - AMOUNT )`

To me it was a bit confusing that it says "given reserve of selling asset and reserve of buying asset".
One could argue that the three parameters are the given part:
{sell_reserve, buy_reserve, amount}

Do you see any potential issue here that is worth thinking about?

PS:
Sorry for the awful formatting, my formatting skills here are weak :)

File: src/amm.rs [hunk 1]
```diff
@@ -134,8 +133,7 @@ pub fn calculate_liquidity_in(asset_a_reserve: Balance, asset_b_reserve: Balance
     to_u128!(b_required_hp)
 }
 
-// TODO what is being calculated?
-/// Calculating ?
+/// Calculating amount of assets returned when removing liquidity.
```
Comment (RoboRambo) [inline src/amm.rs]: @enthusiastmartin Is this too concrete? -> "when removing liquidity"

it might be if this is function is to be used in a wider context, but afaik this is the only context at the moment and the usability of this function outside of this context is probably questionable.

**Comment by enthusiastmartin** [comment]
> Docs for fn calculate_buy_price state:
> `/// Calculating buying price given reserve of selling asset and reserve of buying asset. /// Formula : SELL_RESERVE * AMOUNT / (BUY_RESERVE - AMOUNT )`

yes. correct. definitely needs better description. Docs/words are hard.

**Comment by jak-pan** [comment]
> Thanks for the reply, I will surely look into property testing.
> 
> Had a conversation regarding issue today with jakub, need to process all the data :)
> 
> I do have one question though:
> Docs for fn calculate_buy_price state:
> `/// Calculating buying price given reserve of selling asset and reserve of buying asset. /// Formula : SELL_RESERVE * AMOUNT / (BUY_RESERVE - AMOUNT )`
> 
> To me it was a bit confusing that it says "given reserve of selling asset and reserve of buying asset".
> One could argue that the three parameters are the given part:
> {sell_reserve, buy_reserve, amount}
> 
> Do you see any potential issue here that is worth thinking about?
> 
> PS:
> Sorry for the awful formatting, my formatting skills here are weak :)

Yes the docs could use better wording. We also have a movement to get rid of the sell and buy references as it is clear to some people but completely confusing to other. We are going to use in and out like everybody else does instead of sell and buy (change from the point of actor to point of the pool itself) The only place where sell and buy will be left probably is user facing functions. This could clear some stuff. So here it would change to.

reserve_in * amount / (reserve_out - amount) and the corresponding docs should be updated. I'll probably create new issue linking these.

File: src/amm.rs (hunk 1 above)
Comment (jak-pan) [reply src/amm.rs]: It's ok. We'll probably rename this once we have more pool types to something like xyk_math and then it will be clear. This lib is only to be used in a pool scope.

File: src/proptests.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
+use proptest::test_runner::{TestRunner};
+use proptest::strategy::{Strategy, ValueTree};
+use proptest::prelude::*;
+
```
Comment (RoboRambo) [inline src/proptests.rs]: I feel like I am not using the syntax in the right way, feedback would be appreciated.

File: src/proptests.rs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            assert_eq!(val.0, 1);
+            assert!(val.1 >= 33499370958229846944519890329592825);
+            assert!(val.2 >= 33499370958229846944519890329592825);
+        }
+    }
+}
+
+
+/// Trying to find minimal failing cases for #calculate_spot_price
+#[allow(unused)]
```
Comment (RoboRambo) [inline src/proptests.rs]: When trying to find minimal failing cases, we can make the variable sell_reserve constant and try to simplify and complicate buy_amount & buy_reserve, as the formula is:

/// Formula : BUY_RESERVE * AMOUNT / SELL_RESERVE

Hence the upper bound values for buy_reserve & amount are determined by sell_reserve.

It is quite tricky to implement Strategies for this one without duplicating the function logic.
Of course I am not satisfied with this, however we could at least guarantee a correct result if buy_reserve and amount each are lower than 33499370958229846944519890329592825.

File: src/amm.rs [hunk 4]
```diff
@@ -56,6 +56,10 @@ macro_rules! to_u128 {
 pub fn calculate_spot_price(sell_reserve: Balance, buy_reserve: Balance, amount: Balance) -> Option<Balance> {
     ensure!(sell_reserve != 0);
 
+    if amount == 0 || buy_reserve == 0 {
```
Comment (RoboRambo) [inline src/amm.rs]: Here again the formula is:

/// Formula : BUY_RESERVE * AMOUNT / SELL_RESERVE

So if either amount or buy_reserve are zero, there is no need to compute anything.
Does this make sense here?

File: src/amm.rs [hunk 5]
```diff
@@ -75,6 +79,10 @@ pub fn calculate_spot_price(sell_reserve: Balance, buy_reserve: Balance, amount:
 ///
 /// Returns None in case of error
 pub fn calculate_sell_price(sell_reserve: Balance, buy_reserve: Balance, sell_amount: Balance) -> Option<Balance> {
```
Comment (RoboRambo) [inline src/amm.rs]: In #calculate_spot_price we check if the denominator is zero, yet here we dont.
So these functions will create different results of the denominator is zero as can be seen in the tests.

Is this intended?

File: src/tests.rs [hunk 6]
```diff
@@ -16,15 +19,24 @@ fn spot_price_should_work() {
...
 #[test]
 fn sell_price_should_work() {
     let cases = vec![
         (1000, 2000, 500, Some(667), "Easy case"),
-        (0, 0, 0, None, "Zero reserves and weights"),
-        (0, 1, 0, None, "Zero sell reserve and amount"),
-        (1, 0, 0, Some(1), "Zero buy reserve and amount"),
-        (0, 0, u128::MAX, Some(1), "Zero buy reserve and sell reserve"),
         (0, u128::MAX, u128::MAX, None, "Zero sell reserve"),
+
```
Comment (RoboRambo) [inline src/tests.rs]: for #calculate_sell_price the formula is:
/// Formula : BUY_RESERVE * SELL_AMOUNT / (SELL_RESERVE + SELL_AMOUNT )

So if either buy_reserve or sell_amount is zero, the numerator will be zero and again we dont have to compute anyhting.

Now If we include a conditional for these cases, the behaviour of our method changes quite drastically.
Since we round_up our result in the end, the following holds:
If Numerator = (BUY_RESERVE * SELL_AMOUNT) = 0, then we return 1.

I am not sure if this should be the case.

**Comment by enthusiastmartin** [comment]
Thanks. I will go through all these changes shortly.

But can i revert the "crate renamed" and make as separate change in new PR ?  Probably same with the renaming of sell/buy with in/out. So we dont have multiple issues in one PR.

**Comment by RoboRambo** [comment]
Of course, please do so.

I was wondering how you would like these issues to be split up into PRs, but i didnt want to ask more questions as i already asked a zillion :D

I can do it as well, so u can focus on other things

**Comment by RoboRambo** [comment]
Done.

I'll create seperate PRs for these changes.

**Comment by enthusiastmartin** [comment]
> Done.
> 
> I'll create seperate PRs for these changes.

yes, thanks. I really meant "can you revert" in the original posts. so great.
